ipwb/replay.py

```diff
@@ -603,6 +614,9 @@ def handler(signum, frame):
 
     resp.headers['Memento-Datetime'] = ipwbConfig.datetimeToRFC1123(datetime)
 
+    if not headerCatSuccess:
+        resp.headers['X-Headers-Generated-By'] = 'InterPlanetary Wayback'
```

Comment:
In this case, the response will not have a `content-type` and `content-encoding` headers because those headers are extracted and added from the original headers:

https://github.com/oduwsdl/ipwb/blob/61ec6a5f643d2659483d5621c61e88f408b0d197/ipwb/replay.py#L603-L606

I think we need to surface these two essential headers (the second one is optional) in the CDXJ file (of which the `content-type` is already stored as `mime_type` field). Then when generating the response, we don't need this `not in` conditional, instead, use them directly from the index.

Reply:
Content-type comes over, as it's pulled from the CDXJ but the encoding issue remains. The test replay has gremlins if the encoding is not specified.

> I think we need to surface these two essential headers

From where? This ticket addresses the issue where the headers are not available. Are you suggesting we include this information in the CDXJ?

Reply:
What do we do when this information (content type, content-encoding) is not available in the CDXJ?

Reply:
> Content-type comes over, as it's pulled from the CDXJ but the encoding issue remains. The test replay has gremlins if the encoding is not specified.

Although we do store the `mime_type` in the index, we do not use it at the time of replay and use the one in the original headers instead. We need to also store the `content-encoding` field in the index (if present) and at the time of replay we need to use these fields from the index rather than the original headers.

> From where? This ticket addresses the issue where the headers are not available.

They are not absent, but took longer time to fetch from the IPFS. At the time of indexing this will be available (MIME certainly, Encoding possibly).

Reply:
> What do we do when this information (content type, content-encoding) is not available in the CDXJ?

`content-type` is mostly available in responses, `content-encoding` is only present when necessary. So, in the replay use them if they are present, otherwise let the user-agent decide how to deal with it. It is the best we can do, because that's how original resource was served when it was live.

Reply:
Right, ok, so we need to actually use the MIME type from the index on replay.

My question was, "What do we do when they are absent?" as in, if the CDXJ gets stripped down to not have this information. What is the best way to still allow the payload to be replay without this information being present?

I think your suggestions here are good and have opened up #376 and #377 (whac-a-mole!)

Reply:
> My question was, "What do we do when they are absent?" as in, if the CDXJ gets stripped down to not have this information. What is the best way to still allow the payload to be replay without this information being present?

"You get what you get and you don't get upset" - DSHR

Reply:
Should we try a best-guess at the content encoding and content type in this situation?

Reply:
No, let the user-agents make the best-guess, as those are more mature and can handle some of these things way better than our naive approach would.

Reply:
The user agent currently displays the encoding metadata...
<img width="698" alt="screen shot 2018-01-30 at 6 40 34 pm" src="https://user-images.githubusercontent.com/2514780/35597549-1709a38c-05ed-11e8-9963-beb17d87739a.png">

Reply:
If we tried fixing that, we might have ended up in even worse detection. I think once we have this info added to the index, such corner cases will not arise very frequently.